src/v0/destinations/ga4/customMappingsHandler.js

```diff
@@ -0,0 +1,198 @@
+
+    // take optional params parameters for custom events
+    ga4EventPayload.params = {
+      ...ga4EventPayload.params,
+      ...constructPayload(message, mappingConfig[ConfigCategory.TrackPageCommonParamsConfig.name]),
+    };
+
+    // all extra parameters passed is incorporated inside params
+    ga4EventPayload.params = getGA4CustomParameters(
+      message,
+      ['properties'],
+      GA4_RESERVED_PARAMETER_EXCLUSION.concat(GA4_PARAMETERS_EXCLUSION),
+      ga4EventPayload,
+    );
+
+    // Prepare GA4 user properties
+    const userProperties = prepareUserProperties(message, Config.piiPropertiesToIgnore);
+    if (!isEmptyObject(userProperties)) {
+      rawPayload.user_properties = userProperties;
+    }
+
+    rawPayload = removeUndefinedAndNullValues(rawPayload);
+    rawPayload = { ...rawPayload, events: [ga4EventPayload] };
+
+    boilerplateOperations(rawPayload, message, Config);
+    return rawPayload;
+  }
+
+  const processedPayloads = validMappings.map((mapping) => {
+    const eventName = mapping.ga4EventName;
+    // reserved event names are not allowed
+    if (isReservedEventName(eventName)) {
+      throw new InstrumentationError(`[GA4]:: Reserved event name: ${eventName} are not allowed`);
+    }
+    // validation for ga4 event name
+    validateEventName(eventName);
+
+    // Add common top level payload
+    let ga4BasicPayload = constructPayload(message, trackCommonConfig);
+    ga4BasicPayload = addClientDetails(ga4BasicPayload, message, Config);
+
+    const eventPropertiesMappings = mapping.eventProperties || {};
+
+    const ga4MappedPayload = {};
+
+    for (const propertyMapping of eventPropertiesMappings) {
+      mapWithJsonPath(message, ga4MappedPayload, propertyMapping.from, propertyMapping.to);
+    }
+    removeUndefinedAndNullRecurse(ga4MappedPayload);
+
+    boilerplateOperations(ga4MappedPayload, message, Config);
+
+    if (isDefinedAndNotNull(ga4BasicPayload)) {
+      return { ...ga4BasicPayload, ...ga4MappedPayload };
+    } else {
+      return ga4MappedPayload;
+    }
+  });
+
+  return processedPayloads.map((processedPayload) =>
+    buildDeliverablePayload(processedPayload, Config),
+  );
+};
+
+const boilerplateOperations = (ga4Payload, message, Config) => {
+  removeReservedParameterPrefixNames(ga4Payload.events[0].params);
+
+  const integrationsObj = getIntegrationsObj(message, 'ga4');
+
+  if (isHybridModeEnabled(Config) && integrationsObj && integrationsObj.sessionId) {
+    ga4Payload.events[0].params.session_id = integrationsObj.sessionId;
+  }
+
+  if (ga4Payload.events[0].params) {
+    ga4Payload.events[0].params = removeInvalidParams(
+      removeUndefinedAndNullValues(ga4Payload.events[0].params),
+    );
+  }
+
+  if (isEmptyObject(ga4Payload.events[0].params)) {
+    delete ga4Payload.events[0].params;
+  }
+
+  // Prepare GA4 consents
+  const consents = prepareUserConsents(message);
+  if (!isEmptyObject(consents)) {
+    ga4Payload.consent = consents;
+  }
+};
+
```

Comment:
Move this function to new util class under GA4 itself - to make this class file more readable.

Reply:
Addressed